students/eowyn/backcast.py

```diff
@@ -0,0 +1,252 @@
+    Drop all Feb 29 rows:
+    temp2 = temp.drop(temp.index[1416:1440], axis = 0)
+    """
+    if len(met_yr.index) == 8784:
+        print("removing leap day from met year")
+        met_yr = met_yr.drop(met_yr.index[1416:1440], axis=0)
+    if len(pwr_yr.index) == 8784:
+        print("removing leap day from power year")
+        pwr_yr = pwr_yr.drop(pwr_yr.index[1416:1440], axis=0)
+
+    # Construct a dummy TimeStamp index for the typical year
+    typical_year = pd.date_range('2015-01-01', periods=8760, freq='H')
+    met_yr.index, met_yr.index.names = typical_year, ["TimeStamp"]
+    pwr_yr.index, pwr_yr.index.names = typical_year, ["TimeStamp"]
+
+    # Concatenate the data together to form a single data frame
+    aligned_data = pd.concat([met_yr, pwr_yr],axis=1)
+
+"""
+In [168]: y = pd.date_range('2015-08-03', periods=1000, freq='H')
+use that to help write tests
+
+temp = met_hour.truncate('2009-01-01 00:00:00','2009-12-31 23:00:00').index
+
+"""
+
+"""
+Determine profit for a given substation, in this case, Brown.
+Add Profit column to aligned_data
+TO DO: Make choice of substation (or ave of all stations) an option
+"""
+aligned_data["Revenue"] = 1e-4*aligned_data["Generation"]*aligned_data["Brown"]
+
+"""
+Generate monthly averages of peak and off-peak revenue
+Print tables
+TO DO: Take args for peak/off-peak hours
+"""
+
+# Filter aligned_data into peakdf and offpeakdf
+peakHours = pd.Series([i for i in range(0, 12)])  # 0..11
+offPeakHours = pd.Series([i for i in range(12, 24)])  # 12..23
+peakTimes = pd.Series(aligned_data.index.hour).isin(peakHours)  # Boolean
+peakTimes.index, peakTimes.index.names = typical_year, ["TimeStamp"]
+offPeakTimes = pd.Series(aligned_data.index.hour).isin(offPeakHours)  # Boolean
+offPeakTimes.index, offPeakTimes.index.names = typical_year, ["TimeStamp"]
+peakTimes = peakTimes.apply(lambda x: 1 if x else np.nan)
+offPeakTimes = offPeakTimes.apply(lambda x: 1 if x else np.nan)
+
+
+peakDF = aligned_data.mul(peakTimes, axis=0)
+peakDF = peakDF.drop(['North','West','South','Red Creek','Brown'],axis=1)
+offPeakDF = aligned_data.mul(offPeakTimes, axis=0)
+offPeakDF = offPeakDF.drop(['North','West','South','Red Creek','Brown'],axis=1)
+
+# Finally, pretty print tables for peak and off mean monthly means.
+# Done: group wind by MEAN and generation, revenue by SUM
+# Done: Write results to file with full precision
+windname = peakDF.columns[0]  
+print(60*'-')
+print('Peak Hours Monthly Sum')
+result = peakDF.groupby(peakDF.index.month).agg({windname: np.mean, "Generation": sum, "Revenue":sum})
+result.index.names=["Month"]
+print(tabulate(result,headers = 'keys'))
+result.to_csv('PeakHours-Revenue.txt', sep='\t')
+print(60*'-')
+print('Off Peak Hours Monthly Sum')
+result = offPeakDF.groupby(offPeakDF.index.month).agg({windname: np.mean, "Generation": sum, "Revenue":sum})
+result.index.names=["Month"]
+print(tabulate(result,headers = 'keys'))
+result.to_csv('OffPeakHours-Revenue.txt', sep='\t')
+
+
+
+
+
```

Comment:
you probably want to collect all the function in one (or more) modules, and then have a main script (or main function) that  runs through the procedure and calls those functions.

And probably group them into "clean up the data" and "do the actual analysis"